sql/parser2/eval.go

```diff
@@ -20,31 +20,308 @@ package parser2
 import (
 	"fmt"
 	"math"
+	"reflect"
 	"strconv"
-
-	"github.com/cockroachdb/cockroach/sql/sqlwire"
 )
 
-type opType int
+// TODO(pmattis):
+//
+// - Support tuples (i.e. []Datum) and tuple operations (a IN (b, c, d)).
+//
+// - Support decimal arithmetic.
+
+// A Datum holds either a bool, int64, float64, string or []Datum.
+type Datum interface {
+	IsNull() bool
+	ToBool() (Datum, error)
+	String() string
+}
+
+var _ Datum = dbool(false)
+var _ Datum = dint(0)
+var _ Datum = dfloat(0)
+var _ Datum = dstring("")
+var _ Datum = dnull{}
+
+type dbool bool
+
+func (d dbool) IsNull() bool {
+	return false
+}
+
+func (d dbool) ToBool() (Datum, error) {
+	return d, nil
+}
+
+func (d dbool) String() string {
+	if d {
+		return "true"
+	}
+	return "false"
+}
+
+type dint int64
+
+func (d dint) IsNull() bool {
+	return false
+}
+
+func (d dint) ToBool() (Datum, error) {
+	return dbool(d != 0), nil
+}
+
+func (d dint) String() string {
+	return strconv.FormatInt(int64(d), 10)
+}
+
+type dfloat float64
 
-const (
-	intOp opType = iota
-	uintOp
-	floatOp
-	stringOp
+func (d dfloat) IsNull() bool {
+	return false
+}
+
+func (d dfloat) ToBool() (Datum, error) {
+	return dbool(d != 0), nil
+}
+
+func (d dfloat) String() string {
+	return strconv.FormatFloat(float64(d), 'g', -1, 64)
+}
+
+type dstring string
+
+func (d dstring) IsNull() bool {
+	return false
+}
+
+func (d dstring) ToBool() (Datum, error) {
+	return null, fmt.Errorf("cannot convert string to bool")
+}
+
+func (d dstring) String() string {
+	return string(d)
+}
+
+type dnull struct{}
+
+func (d dnull) IsNull() bool {
+	return true
+}
+
+func (d dnull) ToBool() (Datum, error) {
+	return d, fmt.Errorf("cannot convert NULL to bool")
+}
+
+func (d dnull) String() string {
+	return "NULL"
+}
+
+var null = dnull{}
```

Comment:
Compiler complains that `const initializer dnull literal is not a constant`. I think constants can only be basic types.